SMLHelper/Options/ConfigModOptions.cs

```diff
@@ -0,0 +1,952 @@
+
+                // If there is no label specified, just use the member's name.
+                if (string.IsNullOrEmpty(modOptionAttribute.Label))
+                    modOptionAttribute.Label = memberInfo.Name;
+
+                // ModOptionMetadata needed for all ModOptions
+                var modOptionMetadata = new ModOptionMetadata
+                {
+                    ModOptionAttribute = modOptionAttribute,
+                    MemberInfoMetadata = new MemberInfoMetadata
+                    {
+                        MemberType = memberType,
+                        Name = memberInfo.Name,
+                        ValueType = underlyingType
+                    },
+                    OnGameObjectCreatedMetadata = GetEventMetadata<OnGameObjectCreatedAttribute>(memberInfo)
+                };
+
+                if (memberType == MemberType.Method)
+                    modOptionMetadata.MemberInfoMetadata.ParseMethodParameterTypes(memberInfo as MethodInfo);
+
+                if (typeof(TAttribute) != typeof(ButtonAttribute))
+                    modOptionMetadata.OnChangeMetadata = GetEventMetadata<OnChangeAttribute>(memberInfo);
+
+                modOptionsMetadata.Add(modOptionAttribute.Id, modOptionMetadata);
+            }
+            catch (Exception ex)
+            {
+                Logger.Error($"[ConfigModOptions] {ex.Message}");
+            }
+        }
+
+        /// <summary>
+        /// Gets the metadata of every <typeparamref name="TAttribute"/> defined for a member.
+        /// </summary>
+        /// <typeparam name="TAttribute">
+        /// The type of <see cref="ModOptionEventAttribute"/> attribute defined on the member to gather metadata for.
+        /// </typeparam>
+        /// <param name="memberInfo">The member to gather attribute metadata for.</param>
+        /// <returns></returns>
+        private IEnumerable<MemberInfoMetadata> GetEventMetadata<TAttribute>(MemberInfo memberInfo)
+            where TAttribute : ModOptionEventAttribute
+        {
+            var metadatas = new List<MemberInfoMetadata>();
+            foreach (TAttribute attribute in memberInfo.GetCustomAttributes(typeof(TAttribute), true))
+            {
+                var methodMetadata = new MemberInfoMetadata
+                {
+                    MemberType = MemberType.Method,
+                    Name = attribute.MethodName
+                };
+                methodMetadata.ParseMethodParameterTypes();
+                metadatas.Add(methodMetadata);
+            }
+            return metadatas;
+        }
+        #endregion
+
+        #region Click, Change and GameObjectCreated Events
+        /// <summary>
+        /// Conditionally binds events, ie. OnButtonClicked, OnKeybindChanged, etc.
+        /// </summary>
+        private void BindEvents()
+        {
+            if (modOptionsMetadata.Values.Any(x => x.ModOptionAttribute is ButtonAttribute))
```

Comment:
Multiple empty events being called over and over and over again for every mod using this feature for no reason is heavier than a conditional check that happens once per mod.

Reply:
Those events are separate for each `ModOptions` instance, if you don't have particular option type in your config, event for that option type will not be raised for that instance.

Reply:
`ConfigModOptions<T> where T : ConfigFile, new()` adds an event listener *conditionally* for each of those events, if `T` has a need for them. If we don't conditionally add the event listener, then whether you do or don't have that specific option type in your config, our internal event will be raised to loop through your applicable options, empty or not, to raise it. And `ConfigModOptions<T>` event will be raised whether or not you have any. Which, if there are none, is *more* wasteful than checking before we add the listener.

Reply:
Say we didn't check, and we added the event listeners on `ConfigModOptions<T>` like this:

```cs
private void BindEvents() {
  ButtonClicked += ConfigModOptions_ButtonClicked;
  ChoiceChanged += ConfigModOptions_ChoiceChanged;
  KeybindChanged += ConfigModOptions_KeybindChanged;
  SliderChanged += ConfigModOptions_SliderChanged;
  ToggleChanged += ConfigModOptions_ToggleChanged;
  GameObjectCreated += ConfigModOptions_GameObjectCreated;
}

private void ConfigModOptions_ChoiceChanged(object sender, ChoiceChangedEventArgs e)
{
  if (modOptionsMetadata.TryGetValue(e.Id, out var modOptionMetadata)
      && modOptionMetadata.MemberInfoMetadata.MethodParameterTypes is Type[] parameterTypes)
  {
      // do stuff
  }
}
// etc.
```

Then, any time any choice is changed, our ChoiceChanged event is fired up, and needlessly checks the dictionary for a value we already know won't be there. Every time. It is clearly therefore better to just not listen to the event if it's not needed.

Reply:
Hmm, I completely lost, is it possible to add option to the same `ModOptions` instance that handled by `ConfigFile` bypassing `ConfigFile` ?

Reply:
Are you asking if its possible to add a modoption manually to the `ConfigModOptions<T>` instance? The answer is no, not through the publicly exposed API, because you never interact with the `ConfigModOptions<T> : ModOptions` instance directly, therefore you don't have access to the `ModOptions` methods for it. Furthermore, it is an `internal` class, so it also cannot be be publicly extended.

If you *did* have access to the `ModOptions` instance, even then you can't add an option to it from outside of it, as `AddChoiceOption` methods etc. are all `protected`.

So no, it is not possible to at runtime add more options to the menu. Because of this, once you register it to the menu, we already know everything we need to know, so we know whether or not we need to listen to the events.

Reply:
>  any time any choice is changed in the menu for T, our ChoiceChanged event is fired up

Ok, so how `ChoiceChanged` would be fired up if you didn't add choice option with `ProcessMetadata()`?

Reply:
Ah, I see your logic now, and that part is fair enough. Still don't see the logic in subscribing to an event we don't need to subscribe to.

Reply:
> Still don't see the logic in subscribing to an event

To avoid running that LINQs and also less code -> more readable.

Reply:
If you want to avoid running LINQ then I could easily write a method to do it without the LINQ expression.

Less code is often (but not always) more readable, but it is not always better. This is getting subjective again.

Reply:
That code that you wrote above (without checks) is easier to read and understand and it's also faster. And I think subscribing to the event that will never fire is a small price for that.

Reply:
@PrimeSonic Can I get a second opinion

Reply:
I haven't had enough time (or mental energy) to fully take this all in.  
So I can't offer any meaningful feedback right now.  
After we go over it together in more detail, I may have something useful to contribute.

Reply:
Resolved by 5259a2e